Feature: ERC721 NFT Contract

0xNeshi opened the pull request:
**Issue(s): Close #197**

### Description

Implement a simple ERC721 contract inspired by OZ's implementation.

_Note: should replace non-maintained PR https://github.com/NethermindEth/StarknetByExample/pull/214_

### Checklist

- [x] **CI Verifier:** Run `./scripts/cairo_programs_verifier.sh` successfully
- [x] **Contract Tests:** Added tests to cover the changes
- [x] **Deployed Contract (Optional):** https://sepolia.voyager.online/contract/0x3e20bc0d30cd99635f27fd2af1e4cc23fc21e87b5f4101c02b41a8f62e83e8a#code

julio4 commented:
@0xNeshi The repository has been upgraded to a new site generator and the Markdown format to include code listings is slightly different. The [CONTRIBUTING](https://github.com/NethermindEth/StarknetByExample/blob/main/CONTRIBUTING.md) guidelines has been updated to reflect these changes.
Could you rebase this PR onto `dev`. If there's any issues or if you want me to do these changes let me know.

julio4 commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,193 @@
+            self.operator_approvals.read((owner, operator))
+        }
+
+        fn mint(ref self: ContractState, to: ContractAddress, token_id: u256) {
+            assert(!to.is_zero(), Errors::INVALID_RECEIVER);
+            assert(self.owners.read(token_id).is_zero(), Errors::ALREADY_MINTED);
+
+            self.balances.write(to, self.balances.read(to) + 1);
+            self.owners.write(token_id, to);
+
+            self.emit(Transfer { from: Zero::zero(), to, token_id });
+        }
+
+        fn burn(ref self: ContractState, token_id: u256) {
+            let owner = self._require_owned(token_id);
+
+            self.balances.write(owner, self.balances.read(owner) - 1);
+
+            self.owners.write(token_id, Zero::zero());
+            self.approvals.write(token_id, Zero::zero());
+
+            self.emit(Transfer { from: owner, to: Zero::zero(), token_id });
+        }
```
Move these in the internal impl as `_mint` and `_burn`, and implement `mint` and `burn` in a separate interface impl block that call the internal functions

julio4 reviewed (commented):
LGTM, a few small changes to follow the IERC721 as much as possible!

0xNeshi commented:
Thanks for the review!
Will address all change requests as I get some spare time this week

0xNeshi commented on src/applications/erc721.md:
````diff
@@ -0,0 +1,19 @@
+{{#include ../../listings/applications/721/src/erc721.cairo}}
+```
+
+There are other implementations, such as the [Open Zeppelin](https://docs.openzeppelin.com/contracts-cairo/0.7.0/erc721) one.
````
Updated to 0.20.0 as the most recent one, see https://docs.openzeppelin.com/contracts-cairo/0.20.0/erc721

Updated the link for erc20 too, see https://github.com/NethermindEth/StarknetByExample/pull/250/commits/4e9607c6bfa76bba89bc43cfa4f1354f9b20c50c

0xNeshi replied:
Removed version (link still works), see https://github.com/NethermindEth/StarknetByExample/pull/250/commits/e3d04fa1b8dcc40ce9454c95646d05996d287af9

0xNeshi replied:
@julio4 removed version (link still works), see https://github.com/NethermindEth/StarknetByExample/pull/250/commits/e3d04fa1b8dcc40ce9454c95646d05996d287af9

0xNeshi commented on src/applications/erc721.md:
```diff
@@ -0,0 +1,19 @@
+# ERC721 Token
+
+Contracts that follow the [ERC721 Standard](https://eips.ethereum.org/EIPS/eip-721) are called ERC721 tokens. They are used to represent non-fungible assets.
```
https://github.com/NethermindEth/StarknetByExample/pull/250/commits/b92ae564a69aa450bb7d95de8b6faa4587d4ef18

0xNeshi commented on listings/applications/erc721/src/interfaces.cairo:
Only the interfaces, no implementation?

0xNeshi commented:
@julio4 addressed all the comments.
Note that the [previously deployed contract](https://sepolia.voyager.online/contract/0x02232292c88e0f69573bcb139c55deeda4f449a85a9fc9c26909f27a4d3f619a#code) is now a bit out-of-sync with the current implementation, although it works the same functionality-wise (`burn` and `mint` are part of separate implementations, but the contract ABI is the same).

0xNeshi commented on pnpm-lock.yaml:
```diff
@@ -1384,6 +1384,9 @@ packages:
   emoji-regex-xs@1.0.0:
     resolution: {integrity: sha512-LRlerrMYoIDrT6jgpeZ2YYl/L8EulRTt5hQcYjy5AInh7HWXKimpqx68aknBFpGL2+/IcogTcaydJEgaTmOpDg==}
 
+  emoji-regex-xs@1.0.0:
+    resolution: {integrity: sha512-LRlerrMYoIDrT6jgpeZ2YYl/L8EulRTt5hQcYjy5AInh7HWXKimpqx68aknBFpGL2+/IcogTcaydJEgaTmOpDg==}
+
```
```suggestion

```

0xNeshi commented on pnpm-lock.yaml:
```diff
@@ -3493,9 +3496,6 @@ snapshots:
   '@rollup/rollup-win32-ia32-msvc@4.34.6':
     optional: true
 
```
```suggestion

  '@rollup/rollup-win32-x64-msvc@4.34.6':
    optional: true

```

0xNeshi commented on pages/applications/erc721.md:
````diff
@@ -0,0 +1,23 @@
+// [!include ~/listings/applications/erc721/src/erc721.cairo]
+```
+
+There are other implementations, such as the [Open Zeppelin](https://docs.openzeppelin.com/contracts-cairo/0.20.0/erc721) one.
````
```suggestion
There are other implementations, such as the [Open Zeppelin](https://docs.openzeppelin.com/contracts-cairo/erc721) one.
```

0xNeshi commented on pages/applications/erc20.md:
````diff
@@ -17,4 +17,4 @@ Here's an implementation of the ERC20 interface in Cairo:
 // [!include ~/listings/applications/erc20/src/token.cairo:erc20]
 ```
 
-There's several other implementations, such as the [Open Zeppelin](https://docs.openzeppelin.com/contracts-cairo/1.0.0/erc20) or the [Cairo By Example](https://cairo-by-example.com/examples/erc20/) ones.
+There's several other implementations, such as the [Open Zeppelin](https://docs.openzeppelin.com/contracts-cairo/erc20) or the [Cairo By Example](https://cairo-by-example.com/examples/erc20/) ones.
````
Makes erc20 cairo link version-agnostic

julio4 reviewed (approved):
LGTM

julio4 commented on listings/applications/erc721/Scarb.toml:
```diff
@@ -0,0 +1,17 @@
+
+[dependencies]
+starknet.workspace = true
+openzeppelin.workspace = true
```
Need to change to openzeppelin packages, i.e. `openzeppelin_introspection`

0xNeshi replied:
```suggestion
openzeppelin_account.workspace = true
openzeppelin_introspection.workspace = true
```

0xNeshi commented on listings/applications/erc721/src/mocks/account.cairo:
```diff
@@ -0,0 +1,46 @@
+        #[substorage(v0)]
+        pub account: AccountComponent::Storage,
+        #[substorage(v0)]
+        pub src5: SRC5Component::Storage
```
```suggestion
        pub src5: SRC5Component::Storage,
```

0xNeshi commented on listings/applications/erc721/src/mocks/account.cairo:
```diff
@@ -0,0 +1,46 @@
+        #[flat]
+        AccountEvent: AccountComponent::Event,
+        #[flat]
+        SRC5Event: SRC5Component::Event
```
```suggestion
        SRC5Event: SRC5Component::Event,
```

0xNeshi commented on listings/applications/erc721/src/interfaces.cairo:
```diff
@@ -0,0 +1,74 @@
+        from: ContractAddress,
+        to: ContractAddress,
+        token_id: u256,
+        data: Span<felt252>
```
```suggestion
        data: Span<felt252>,
```

0xNeshi commented on listings/applications/erc721/src/interfaces.cairo:
```diff
@@ -0,0 +1,74 @@
+    fn set_approval_for_all(ref self: TContractState, operator: ContractAddress, approved: bool);
+    fn get_approved(self: @TContractState, token_id: u256) -> ContractAddress;
+    fn is_approved_for_all(
+        self: @TContractState, owner: ContractAddress, operator: ContractAddress
```
```suggestion
        self: @TContractState, owner: ContractAddress, operator: ContractAddress,
```

0xNeshi commented on listings/applications/erc721/src/interfaces.cairo:
```diff
@@ -0,0 +1,74 @@
+        operator: ContractAddress,
+        from: ContractAddress,
+        token_id: u256,
+        data: Span<felt252>
```
```suggestion
        data: Span<felt252>,
```

0xNeshi commented on listings/applications/erc721/src/mocks/receiver.cairo:
```diff
@@ -0,0 +1,50 @@
+    #[storage]
+    pub struct Storage {
+        #[substorage(v0)]
+        pub src5: SRC5Component::Storage
```
```suggestion
        pub src5: SRC5Component::Storage,
```

0xNeshi commented on listings/applications/erc721/src/mocks/receiver.cairo:
```diff
@@ -0,0 +1,50 @@
+            operator: ContractAddress,
+            from: ContractAddress,
+            token_id: u256,
+            data: Span<felt252>
```
```suggestion
            data: Span<felt252>,
```

0xNeshi commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,208 @@
+    pub struct Transfer {
+        pub from: ContractAddress,
+        pub to: ContractAddress,
+        pub token_id: u256
```
```suggestion
        pub token_id: u256,
```

0xNeshi commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    #[derive(Drop, starknet::Event)]
+    pub struct Approval {
+        pub owner: ContractAddress,
+        pub approved: ContractAddress,
+        pub token_id: u256
+    }
+
+    #[derive(Drop, starknet::Event)]
+    pub struct ApprovalForAll {
+        pub owner: ContractAddress,
+        pub operator: ContractAddress,
+        pub approved: bool
+    }
```
```suggestion
    pub struct Approval {
        pub owner: ContractAddress,
        pub approved: ContractAddress,
        pub token_id: u256,
    }

    #[derive(Drop, starknet::Event)]
    pub struct ApprovalForAll {
        pub owner: ContractAddress,
        pub operator: ContractAddress,
        pub approved: bool,
    }
```

0xNeshi commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,208 @@
+        }
+
+        fn set_approval_for_all(
+            ref self: ContractState, operator: ContractAddress, approved: bool
```
```suggestion
            ref self: ContractState, operator: ContractAddress, approved: bool,
```

0xNeshi commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,208 @@
+            let owner = self._require_owned(token_id);
+            let caller = get_caller_address();
+            assert(
+                caller == owner || self.is_approved_for_all(owner, caller), Errors::UNAUTHORIZED
```
```suggestion
                caller == owner || self.is_approved_for_all(owner, caller), Errors::UNAUTHORIZED,
```

0xNeshi commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,208 @@
+        }
+
+        fn transfer_from(
+            ref self: ContractState, from: ContractAddress, to: ContractAddress, token_id: u256
```
```suggestion
            ref self: ContractState, from: ContractAddress, to: ContractAddress, token_id: u256,
```

0xNeshi commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,208 @@
+            from: ContractAddress,
+            to: ContractAddress,
+            token_id: u256,
+            data: Span<felt252>
```
```suggestion
            data: Span<felt252>,
```

0xNeshi commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,208 @@
+            Self::transfer_from(ref self, from, to, token_id);
+            assert(
+                self._check_on_erc721_received(from, to, token_id, data),
+                Errors::SAFE_TRANSFER_FAILED
```
```suggestion
                Errors::SAFE_TRANSFER_FAILED,
```

0xNeshi commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,208 @@
+        }
+
+        fn is_approved_for_all(
+            self: @ContractState, owner: ContractAddress, operator: ContractAddress
```
```suggestion
            self: @ContractState, owner: ContractAddress, operator: ContractAddress,
```

0xNeshi commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,208 @@
+        }
+
+        fn _is_approved_or_owner(
+            self: @ContractState, owner: ContractAddress, spender: ContractAddress, token_id: u256
```
```suggestion
            self: @ContractState, owner: ContractAddress, spender: ContractAddress, token_id: u256,
```

0xNeshi commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,208 @@
+            from: ContractAddress,
+            to: ContractAddress,
+            token_id: u256,
+            data: Span<felt252>
```
```suggestion
            data: Span<felt252>,
```

0xNeshi commented on listings/applications/erc721/src/erc721.cairo:
```diff
@@ -0,0 +1,208 @@
+            if src5_dispatcher.supports_interface(IERC721_RECEIVER_ID) {
+                IERC721ReceiverDispatcher { contract_address: to }
+                    .on_erc721_received(
+                        get_caller_address(), from, token_id, data
```
```suggestion
                        get_caller_address(), from, token_id, data,
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+use erc721::interfaces::{
+    IERC721, IERC721Dispatcher, IERC721DispatcherTrait, IERC721SafeDispatcher,
+    IERC721SafeDispatcherTrait, IERC721Mintable, IERC721MintableDispatcher,
+    IERC721MintableDispatcherTrait, IERC721BurnableDispatcher, IERC721BurnableDispatcherTrait
```
```suggestion
    IERC721MintableDispatcherTrait, IERC721BurnableDispatcher, IERC721BurnableDispatcherTrait,
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+    IERC721SafeDispatcherTrait, IERC721Mintable, IERC721MintableDispatcher,
+    IERC721MintableDispatcherTrait, IERC721BurnableDispatcher, IERC721BurnableDispatcherTrait
+};
+use erc721::erc721::{ERC721, ERC721::{Event, Transfer, Approval, ApprovalForAll, InternalTrait},};
```
```suggestion
use erc721::erc721::{ERC721, ERC721::{Event, Transfer, Approval, ApprovalForAll, InternalTrait}};
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+    declare, test_address, DeclareResultTrait, ContractClassTrait, start_cheat_caller_address,
+    spy_events, EventSpyAssertionsTrait,
+};
+use starknet::{ContractAddress, contract_address_const,};
```
```suggestion
use starknet::{ContractAddress, contract_address_const};
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+    start_cheat_caller_address(contract_address, OWNER());
+    contract.approve(SPENDER(), TOKEN_ID);
+
+    spy
+        .assert_emitted(
+            @array![
+                (
+                    contract_address,
+                    Event::Approval(
+                        Approval { owner: OWNER(), approved: SPENDER(), token_id: TOKEN_ID }
+                    )
+                )
+            ]
+        );
```
```suggestion
    spy
        .assert_emitted(
            @array![
                (
                    contract_address,
                    Event::Approval(
                        Approval { owner: OWNER(), approved: SPENDER(), token_id: TOKEN_ID },
                    ),
                ),
            ],
        );
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+
+    start_cheat_caller_address(contract_address, OPERATOR());
+    contract.approve(SPENDER(), TOKEN_ID);
+    spy
+        .assert_emitted(
+            @array![
+                (
+                    contract_address,
+                    Event::Approval(
+                        Approval { owner: OWNER(), approved: SPENDER(), token_id: TOKEN_ID }
+                    )
+                )
+            ]
+        );
```
```suggestion
    spy
        .assert_emitted(
            @array![
                (
                    contract_address,
                    Event::Approval(
                        Approval { owner: OWNER(), approved: SPENDER(), token_id: TOKEN_ID },
                    ),
                ),
            ],
        );
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+
+    start_cheat_caller_address(contract_address, owner);
+    contract.transfer_from(owner, owner, TOKEN_ID);
+    spy
+        .assert_emitted(
+            @array![
+                (
+                    contract_address,
+                    Event::Transfer(Transfer { from: owner, to: owner, token_id: TOKEN_ID })
+                )
+            ]
+        );
```
```suggestion
    spy
        .assert_emitted(
            @array![
                (
                    contract_address,
                    Event::Transfer(Transfer { from: owner, to: owner, token_id: TOKEN_ID }),
                ),
            ],
        );
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+
+    start_cheat_caller_address(contract_address, OPERATOR());
+    contract.transfer_from(owner, recipient, token_id);
+    spy
+        .assert_emitted(
+            @array![
+                (
+                    contract_address,
+                    Event::Transfer(Transfer { from: owner, to: recipient, token_id })
+                )
+            ]
```
```suggestion
    spy
        .assert_emitted(
            @array![
                (
                    contract_address,
                    Event::Transfer(Transfer { from: owner, to: recipient, token_id }),
                ),
            ],
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+            @array![
+                (
+                    contract_address,
+                    Event::Transfer(Transfer { from: owner, to: account, token_id: TOKEN_ID })
+                )
+            ]
```
```suggestion
                    Event::Transfer(Transfer { from: owner, to: account, token_id: TOKEN_ID }),
                ),
            ],
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+            @array![
+                (
+                    contract_address,
+                    Event::Transfer(Transfer { from: owner, to: receiver, token_id: TOKEN_ID })
+                )
+            ]
```
```suggestion
                    Event::Transfer(Transfer { from: owner, to: receiver, token_id: TOKEN_ID }),
                ),
            ],
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+}
+
+#[test]
+#[should_panic(expected: ('ENTRYPOINT_NOT_FOUND', 'ENTRYPOINT_FAILED',))]
```
```suggestion
#[should_panic(expected: ('ENTRYPOINT_NOT_FOUND', 'ENTRYPOINT_FAILED'))]
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+            @array![
+                (
+                    contract_address,
+                    Event::Transfer(Transfer { from: owner, to: owner, token_id: TOKEN_ID })
+                )
+            ]
```
```suggestion
                    Event::Transfer(Transfer { from: owner, to: owner, token_id: TOKEN_ID }),
                ),
            ],
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+            @array![
+                (
+                    contract_address,
+                    Event::Transfer(Transfer { from: owner, to: receiver, token_id: TOKEN_ID })
+                )
+            ]
```
```suggestion
                    Event::Transfer(Transfer { from: owner, to: receiver, token_id: TOKEN_ID }),
                ),
            ],
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+            @array![
+                (
+                    contract_address,
+                    Event::Transfer(Transfer { from: owner, to: receiver, token_id: TOKEN_ID })
+                )
+            ]
```
```suggestion
                    Event::Transfer(Transfer { from: owner, to: receiver, token_id: TOKEN_ID }),
                ),
            ],
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+            @array![
+                (
+                    contract_address,
+                    Event::Transfer(Transfer { from: OWNER(), to: ZERO(), token_id: TOKEN_ID })
+                )
+            ]
```
```suggestion
                    Event::Transfer(Transfer { from: OWNER(), to: ZERO(), token_id: TOKEN_ID }),
                ),
            ],
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+        Result::Ok(_) => panic!("`owner_of` did not panic"),
+        Result::Err(panic_data) => {
+            assert(*panic_data.at(0) == 'ERC721: invalid token ID', *panic_data.at(0));
+        }
```
```suggestion
        },
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+        Result::Ok(_) => panic!("`get_approved` did not panic"),
+        Result::Err(panic_data) => {
+            assert(*panic_data.at(0) == 'ERC721: invalid token ID', *panic_data.at(0));
+        }
```
```suggestion
        },
```

0xNeshi commented on listings/applications/erc721/src/tests.cairo:
```diff
@@ -0,0 +1,803 @@
+//
+
+fn assert_state_before_transfer(
+    contract: IERC721Dispatcher, owner: ContractAddress, recipient: ContractAddress, token_id: u256
+) {
+    assert_eq!(contract.owner_of(token_id), owner);
+    assert_eq!(contract.balance_of(owner), 1);
+    assert!(contract.balance_of(recipient).is_zero());
+}
+
+fn assert_state_after_transfer(
+    contract: IERC721Dispatcher, owner: ContractAddress, recipient: ContractAddress, token_id: u256
```
```suggestion
    contract: IERC721Dispatcher, owner: ContractAddress, recipient: ContractAddress, token_id: u256,
) {
    assert_eq!(contract.owner_of(token_id), owner);
    assert_eq!(contract.balance_of(owner), 1);
    assert!(contract.balance_of(recipient).is_zero());
}

fn assert_state_after_transfer(
    contract: IERC721Dispatcher, owner: ContractAddress, recipient: ContractAddress, token_id: u256,
```

0xNeshi commented on Scarb.lock:
```diff
@@ -92,6 +92,14 @@ version = "0.1.0"
+name = "erc721"
+version = "0.1.0"
+dependencies = [
+ "openzeppelin",
```
```suggestion
 "openzeppelin_account",
 "openzeppelin_introspection",
```
